### PR TITLE
Support for Docker metadata/purls, better test support

### PR DESCRIPTION
**Pull request by dpp**
Added support for detecting and parsing Docker metadata.

When a user does `docker save <image_names> -o <target_tar>` Goat Rodeo will now properly parse the metadata and generate package URLs and other information related to ADGs for Docker images.

The `manifest.json` file is parsed and for each `Config` object is tested for the sha256 in the set of files and associated with that file. Docker image sha is based on the sha of the config file.

For each config, all of the layers that the config is associated with are marked `contained:down` so the layers are associated with the config.

Both the JSON for the config and the manifest's information (which contains tags) is added to the `Item`'s metadata.

A Package URL (pURL) is generated for each tag in the manifest for each config.

Tests including tests against an aggregated (multiple images/configs) is included in the test suite.

In order to disambiguate filenames in layers, if an `Item` contains more than 1 filename for an artifact, the gitoids of the containing items are added to the filenames so that it's possible to determine which file is associated with which container. This will make testing a file overlaying another file in different docker image layers possible.

Building ADGs for the optional `repo_ea` collection of files is no longer optional.

As part of the test phase, the `build.sbt` file downloads artifacts from a CDN. In the case of `repo_ea` (which is 5GB+), the file is also untarred. Thus, the initial test run will take a while.

Documentation and the README is updated.

Added support for having the version of Goat Rodeo be defined by the sbt file rather than having the version (as printed) be manually entered.

Added support for a `-V`/`--version` CLI flag and a `-?`/`--help` flag

Updated the file version for emitted Goat Rodeo clusters to 3 because of the change in the way filenames are merged.

### CODE REVIEW

**Review by earldouglas — APPROVED** [review]

**Comment by dpp** [comment]
I've converted this into a draft PR for two reasons:

* Discussions with @CJ-cmd-byte pointed out that the README doesn't have basic running instructions (D'oh!) and I'll update the README.
* @aredridel has an interesting test case of building ADGs of a JAR/POM combo outside a container and then building an ADG of a container that has the JAR/POM combo... the latter ADG should be a superset of the former

File: src/test/scala/DockerSuite.scala [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    val name = "test_data/download/docker_tests/bigtent_2025_03_22_docker.tar"
+
+    val nested = FileWrapper(File(name), name, None)
+    val store1 = ToProcess.buildGraphFromArtifactWrapper(nested)
+
+    val result = store1.purls()
+
+    assertEquals(
+      result,
+      TreeSet("pkg:docker/bigtent@2025_03_22")
```
Comment (aredridel) [inline src/test/scala/DockerSuite.scala:29]: tags can be pretty arbitrary: have you red-teamed what this could be?

File: src/test/scala/DockerSuite.scala [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    val store1 = ToProcess.buildGraphFromArtifactWrapper(nested)
+
+    val result = store1.purls()
+    val expectedpurls = TreeSet(
+      "pkg:docker/postgres@16.6",
+      "pkg:docker/postgres@9.6.12",
+      "pkg:docker/spicelabs%2Fbigtent@0.8.3",
+      "pkg:docker/spicelabs%2Fbigtent@latest",
+      "pkg:docker/spicelabs%2Fgrinder@0.1.0",
+      "pkg:docker/spicelabs%2Fgrinder@latest"
```
Comment (aredridel) [inline src/test/scala/DockerSuite.scala:78]: Oof, %2F. ghcr.io purls are gonna be uuuugly

**Review by aredridel — APPROVED** [review]
I think this looks good. I don't have a firm idea of how the docker layering gets resolved -- a mermaid diagram showing the nodes and arrows would probably help me a lot. It still seems like all the files are still contained by the parent things? Maybe I'm not quite clear on how it works. (or is it forming a lattice, with sibling nodes being the parent of the complete vs final set of files?)

File: src/main/scala/goatrodeo/omnibor/strategies/Docker.scala [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+      }
+
+      val tagBase = Set(tags.map(t => t._1)*)
+
+      val allTags: Seq[(String, String)] = (tags ::: Nil)
+
+      val purls = allTags.map { case (name, label) =>
+        PackageURLBuilder
+          .aPackageURL()
+          .withType("docker")
```
[the remark this answers is not in the document]
Comment (dpp) [reply src/main/scala/goatrodeo/omnibor/strategies/Docker.scala:62]: It's the spec: https://github.com/package-url/purl-spec?tab=readme-ov-file#some-purl-examples

File: src/main/scala/goatrodeo/omnibor/strategies/Docker.scala [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+          .withName(name)
+          .withVersion(label)
```
[the remark this answers is not in the document]
Comment (dpp) [reply src/main/scala/goatrodeo/omnibor/strategies/Docker.scala:64]: Tag == name:label

File: src/test/scala/DockerSuite.scala (hunk 1 above)
Comment (dpp) [reply src/test/scala/DockerSuite.scala:29]: Just pulled the tags from the `manifest.json` file and put it into the pURL. If there's some checking that should be done, I'm down for that... but right now, just copying the data over.

File: src/test/scala/DockerSuite.scala (hunk 2 above)
Comment (dpp) [reply src/test/scala/DockerSuite.scala:78]: Working on cleaning up the Docker pURLs

**Review by dpletter — APPROVED** [review]

File: src/test/scala/DockerSuite.scala (hunk 1 above)
Comment (aredridel) [reply src/test/scala/DockerSuite.scala:29]: Aha, [found where this is defined](https://github.com/distribution/reference/blob/main/reference.go#L18): `/[\w][\w.-]{0,127}/` matches tags.

**Comment by dpp** [comment]
I've updated the code to support WARs and EARs.

The tests are failing because GitHub is blocking the >1GB data download. And even when the data gets downloaded, the JVM OOMs. Not gonna chase those issues for this PR.

**Comment by aredridel** [comment]
Let's comment out the broken tests then and make a ticket to get them fixed

**Comment by dpp** [comment]
The tests are in fact working. They are failing in automation.

There are a couple of failure modes:

* Failure to download or completely download the test files... this is intermittent
* Failure to run the big ADG build which is a JVM heap size issue that I can't figure out how to fix in GitHub Actions

In order to deal with the download issue, we'd have to comment out almost all the test.

In order to deal with the heap size issue, someone with better GitHub action skills than me has to fix the GitHub action to set the heap size of the JVM when invoking `sbt test`

**Comment by dpp** [comment]
Agreed. And broken GitHub actions should not be a bar to accepting code.

Do you want to fix the actions?

**Comment by aredridel** [comment]
Happy to take that on, but I'd argue until this is passing in automation, the bar is completely appropriate. Annoying, for sure! But completely appropriate.

**Comment by aredridel** [comment]
Do you know what resources the tests require to run?

**Comment by dpp** [comment]
The current failure (this seems to be transient) is downloading the test files from the CloudFlare public R2 bucket. See https://github.com/spice-labs-inc/goatrodeo/actions/runs/14472794351/job/40590849848?pr=90#step:5:48

This is part of adding both the large Docker tars and the repo_ea tar (formerly optional on the local machine/tests, but I think it's important to see a non-trivial ADG run as part of the automated tests not just on the local machine.

Thinking about this failure, it's in sbt, not in the actual GitHub action. I'll add retry logic to the download.

The second issue is that building the ADGs OOMs. I don't know how to provision more heap space to sbt in the action... it's around https://github.com/spice-labs-inc/goatrodeo/blob/main/.github/workflows/scala_ci.yml#L36

The alternative (I don't know if this will work) is to pass an env var in and the tests will provision threads for ADG building based on the env var. The more threads Goat Rodeo uses to build an ADG, the more memory.

So... I'll make two changes: retry on the download and looking for `TEST_THREAD_CNT` and if it's set and parses as an Int, use those number of threads.

If you could update the GitHub action to provision more heap to sbt, that'd be super helpful.

Sound like a plan?

**Comment by earldouglas** [comment]
> If you could update the GitHub action to provision more heap to sbt, that'd be super helpful.

You can change this by setting `SBT_OPTS`:

```
export SBT_OPTS="-Xmx4G"
sbt
```

https://www.scala-sbt.org/1.x/docs/Command-Line-Reference.html#sbt+JVM+heap%2C+permgen%2C+and+stack+sizes

**Comment by aredridel** [comment]
Yep! sounds like a plan -- a single retry is a good idea, I wouldn't want to go further than that I think, and it'll reduce the probability of spurious failure quite low.

**Comment by aredridel** [comment]
The actual heap size might be a problem without using larger runners, which we have to pay github for (or provision our own on hetzner boxes): there's only 7GB of RAM available to a standard runner

**Comment by aredridel** [comment]
As a test strategy, I rather agree we should have something non-trivial, but size itself is not the interesting piece (at least in CI) -- if we had somewhat minimized but still interesting data that exercises edge cases, that'd be best long term. A "full" suite that uses a ton more RAM might be something to run occasionally. That said that's a later problem not a now problem.

**Comment by aredridel** [comment]
I think the JVM options were set and taking effect, but the actual task is big: on my machine I measured 26 GB peak RSS and 55 GB peak VSS in the process. 

```
Storage database opened: "measurement.db"
pid:            783237
process id:     180942702983771013
process name:   java
measurement:    180942704048923581
time:           2025-04-16T12:51:16.884

# statm data
size:                                                                                  55 243 968 Ki   // total program size (same as VmSize in /proc/[pid]/status)
resident:                                                                              26 458 372 Ki   // resident set size (same as VmRSS in /proc/[pid]/status)
shared:                                                                                    17 220 Ki   // number of resident shared pages (i.e., backed by a file) (same as RssFile+RssShmem in /proc/[pid]/status)
text:                                                                                           4 Ki   // text (code)
lib:                                                                                            0 Ki   // library (unused since Linux 2.6; always 0)
data:                                                                                  31 590 576 Ki   // data + stack
dt:                                                                                             0 Ki   // dirty pages (unused since Linux 2.6; always 0)

# smaps data (Pss)
thread stacks:                                                                                 36 Ki
heap:                                                                                         100 Ki
anonymous:                                                                             26 425 052 Ki

/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libnio.so       16 516 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/server/classes.jsa       12 310 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/server/libjvm.so        4 691 Ki
/tmp/libzstd-jni-1.5.6-416460919553683157265.so                                               516 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/modules          486 Ki
/usr/lib/x86_64-linux-gnu/libc.so.6                                                            55 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libjava.so           41 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libjli.so           38 Ki
/tmp/hsperfdata_astewart/783237                                                                32 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libzip.so           26 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libverify.so           25 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libnet.so           19 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libjimage.so           19 Ki
/usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2                                                 16 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libjsvml.so           13 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/lib/libextnet.so           10 Ki
/home/astewart/.cache/coursier/arc/https/github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.5%252B11/OpenJDK21U-jdk_x64_linux_hotspot_21.0.5_11.tar.gz/jdk-21.0.5+11/bin/java            9 Ki
/usr/lib/x86_64-linux-gnu/libm.so.6                                                             9 Ki
/usr/lib/x86_64-linux-gnu/librt.so.1                                                            8 Ki
/usr/lib/x86_64-linux-gnu/libdl.so.2                                                            8 Ki

other mappings:                                                                                10 Ki

sum:                                                                                   26 460 045 Ki
```

**Comment by aredridel** [comment]
It'll be about $0.64 per test run on the 16 GB RAM Github instances.

**Review by aredridel — APPROVED** [review]
🎉
